File: backend/ReliefNexus.API/Models/RiskPrediction.cs
namespace ReliefNexus.API.Models;

public class RiskPrediction
{
    public Guid Id { get; set; } = Guid.NewGuid();

    public string Location { get; set; } = string.Empty;
    public double? Latitude { get; set; }
    public double? Longitude { get; set; }

    public double Rainfall1h { get; set; }
    public double Rainfall3h { get; set; }
    public double Rainfall24h { get; set; }
    public double RiverLevel { get; set; }
    public double RiverFlow { get; set; }
    public double Temperature { get; set; }
    public double Humidity { get; set; }
    public double WindSpeed { get; set; }
    public double SoilMoisture { get; set; }
    public double Elevation { get; set; }
    public double PopulationDensity { get; set; }
    public int HistoricalFloodCount { get; set; }
    public double HistoricalSeverity { get; set; }
    public double DrainageCapacity { get; set; }
    public double ForecastRainfall { get; set; }

    public string DisasterType { get; set; } = string.Empty;
    public double RiskScore { get; set; }
    public string RiskLevel { get; set; } = string.Empty;
    public double Confidence { get; set; }

    public string PredictionSource { get; set; } = "Risk Prediction Agent";
    public string ModelVersion { get; set; } = "v1.0";

    public bool RequiresHumanApproval { get; set; }
    public bool IsApproved { get; set; }

    public string ApprovalStatus { get; set; } = "NotRequired";

    public DateTime CreatedAt { get; set; } = DateTime.UtcNow;

    public List<RiskFactor> RiskFactors { get; set; } = new();
}

public class RiskFactor
{
    public Guid Id { get; set; } = Guid.NewGuid();

    public Guid RiskPredictionId { get; set; }

    public string Factor { get; set; } = string.Empty;
    public double Value { get; set; }
    public string Impact { get; set; } = string.Empty;
    public double Contribution { get; set; }

    public RiskPrediction? RiskPrediction { get; set; }
}

File: frontend/src/features/risk-prediction/types/riskPrediction.ts
export type RiskLevel =
  | "Low"
  | "Moderate"
  | "Medium"
  | "High"
  | "Critical"
  | "DataUnavailable"
  | string;

export interface RiskFactor {
  factor: string;
  value: number;
  impact: string;
  contribution: number;
}

export interface RiskPrediction {
  id?: string;

  location: string;
  latitude?: number | null;
  longitude?: number | null;

  rainfall1h: number;
  rainfall3h: number;
  rainfall24h: number;

  riverLevel: number;
  riverFlow: number;

  temperature: number;
  humidity: number;
  windSpeed: number;
  soilMoisture: number;

  elevation: number;
  populationDensity: number;

  historicalFloodCount: number;
  historicalSeverity: number;
  drainageCapacity: number;

  forecastRainfall: number;

  disasterType: string;
  riskScore: number;
  riskLevel: RiskLevel;
  confidence: number;

  riskFactors: RiskFactor[];
  recommendations: string[];

  predictionSource: string;
  modelVersion: string;

  requiresHumanApproval: boolean;
  isApproved: boolean;

  createdAt: string;

  /**
   * Newer versions of the backend risk engine can expose
   * multiple disaster-specific predictions.
   *
   * This is optional so the frontend remains compatible
   * with the currently verified RiskPredictionDto.
   */
  disasterRisks?: DisasterRisk[];
}

export interface DisasterRisk {
  disasterType: string;
  riskScore: number | null;
  riskLevel: RiskLevel;
  dataAvailable?: boolean;
  dataSource?: string;
}

export interface RiskPredictionRequest {
  location: string;

  latitude?: number | null;
  longitude?: number | null;

  rainfall1h: number;
  rainfall3h: number;
  rainfall24h: number;

  riverLevel: number;
  riverFlow: number;

  temperature: number;
  humidity: number;
  windSpeed: number;
  soilMoisture: number;

  elevation: number;
  populationDensity: number;

  historicalFloodCount: number;
  historicalSeverity: number;
  drainageCapacity: number;

  forecastRainfall: number;

  /**
   * The backend calculates these values.
   * They are therefore optional on create.
   */
  disasterType?: string;
  riskScore?: number;
  riskLevel?: string;
  confidence?: number;
}

export interface LocationSearchResult {
  id: number;

  name: string;
  latitude: number;
  longitude: number;

  country?: string;
  countryCode?: string;

  admin1?: string;
  admin2?: string;

  timezone?: string;
  population?: number;
}

export interface OpenMeteoCurrentWeather {
  time?: string;

  temperature_2m?: number;
  relative_humidity_2m?: number;

  precipitation?: number;
  rain?: number;
  showers?: number;

  wind_speed_10m?: number;
  wind_gusts_10m?: number;
  wind_direction_10m?: number;

  weather_code?: number;
}

export interface OpenMeteoHourlyWeather {
  time: string[];

  precipitation?: number[];
  rain?: number[];
  showers?: number[];

  precipitation_probability?: number[];

  temperature_2m?: number[];
  relative_humidity_2m?: number[];

  wind_speed_10m?: number[];
  wind_gusts_10m?: number[];

  soil_moisture_0_to_1cm?: number[];
  soil_moisture_1_to_3cm?: number[];
  soil_moisture_3_to_9cm?: number[];
  soil_moisture_9_to_27cm?: number[];
}

export interface OpenMeteoWeatherResponse {
  latitude: number;
  longitude: number;

  timezone?: string;
  timezone_abbreviation?: string;

  elevation?: number;

  current?: OpenMeteoCurrentWeather;
  hourly?: OpenMeteoHourlyWeather;

  current_units?: Record<string, string>;
  hourly_units?: Record<string, string>;
}

export interface GdacsEvent {
  eventtype?: string;
  eventid?: string | number;
  episodeid?: string | number;

  eventname?: string;

  alertlevel?: string;

  fromdate?: string;
  todate?: string;

  latitude?: number;
  longitude?: number;

  country?: string;
  iso3?: string;

  severity?: number | null;

  [key: string]: unknown;
}

export interface MapMarker {
  id: string;

  latitude: number;
  longitude: number;

  title: string;

  disasterType?: string;
  riskLevel?: RiskLevel;

  source?: string;

  popupData?: Record<string, unknown>;
}

export interface RiskPredictionState {
  prediction: RiskPrediction | null;

  recentPredictions: RiskPrediction[];

  loading: boolean;
  submitting: boolean;

  error: string | null;
}